### PR TITLE
[ECO-914] update `/markets` endpoint to include volume, min, and max

### PR DESCRIPTION
**Pull request by CRBl69**
Added a new aggregator pipeline that is named `UpdateMaterializedView`. This updates the given materialized view every given duration. This will allow easly creating views that can be updated at a set interval with minimal aggregator code.

This PR also adds a materialized view for market volume, min fill, and max fill in the last 24 hours.

### CODE REVIEW

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+use anyhow::anyhow;
+use chrono::{DateTime, Duration, Utc};
+use sqlx::PgPool;
+
+use aggregator::{Pipeline, PipelineAggregationResult, PipelineError};
+
+pub struct UpdateMaterializedView {
```
Comment (alnoki) [inline src/rust/aggregator/src/pipelines/update_materialized_view.rs:7]: Before we lock this into the codebase, how about we go with `RefreshMaterializedView` , to make it clear that we are doing a 1-to-1 PostgreSQL operation?

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    pool: PgPool,
+    view_name: String,
+    update_interval: std::time::Duration,
+    last_indexed_timestamp: Option<DateTime<Utc>>,
+}
+
+#[allow(dead_code)]
```
Comment (alnoki) [inline src/rust/aggregator/src/pipelines/update_materialized_view.rs:14]: Why is this here?

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            update_interval,
+            last_indexed_timestamp: None,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Pipeline for UpdateMaterializedView {
+    fn model_name(&self) -> String {
+        String::from("UpdateMaterializedView")
```
Comment (alnoki) [inline src/rust/aggregator/src/pipelines/update_materialized_view.rs:33]: `RefreshMaterializedView` ?

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    LEFT JOIN
+        first_fills AS f
+    ON
+        f.market_id = m.market_id
+    LEFT JOIN
+        last_fills AS l
+    ON
+        l.market_id = m.market_id
+    LEFT JOIN
+        aggregator.markets_24h_data AS v
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:106]: If the materialized view hasn't been precomputed, will this operation compute it or error out?

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        Some(self.update_interval)
+    }
+
+    async fn process_and_save_internal(&mut self) -> PipelineAggregationResult {
+        sqlx::query(
+            format!(
+                r#"
+                REFRESH MATERIALIZED VIEW {};
+            "#,
+                self.view_name
```
Comment (alnoki) [inline src/rust/aggregator/src/pipelines/update_materialized_view.rs:57]: How about making the raw string a one-liner? It doesn't need three

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    AND
+        r.quote_module_name = m.quote_module_name
+    AND
+        r.quote_struct_name = m.quote_struct_name
+    LEFT JOIN
+        first_fills AS f
+    ON
+        f.market_id = m.market_id
+    LEFT JOIN
+        last_fills AS l
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:102]: I don't see a `LIMIT` statement on either first or last fills

Will this be joining more rows than needed?

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    async fn process_and_save_historical_data(&mut self) -> PipelineAggregationResult {
+        self.process_and_save_internal().await
+    }
+
+    fn poll_interval(&self) -> Option<std::time::Duration> {
+        Some(self.update_interval)
+    }
+
+    async fn process_and_save_internal(&mut self) -> PipelineAggregationResult {
+        sqlx::query(
```
Comment (alnoki) [inline src/rust/aggregator/src/pipelines/update_materialized_view.rs:52]: Do you need to run SQLx prepare for this?

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        m.tick_size,
+        m.min_size,
+        m.underwriter_id,
+        CASE
+            WHEN r.market_id = m.market_id THEN true
+            ELSE false
+        END AS is_recognized,
+        l.price AS last_fill_price_24hr,
+        (COALESCE(l.price, 1) - COALESCE(f.price, 1)) / COALESCE(f.price, 1) * 100 AS percent_change_24h,
+        l.price - f.price AS nominal_change_24h,
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:74]: This is the change in integer prices over 24 hr, whereas prices don't become nominal until coin decimals are accounted for

So how about `AS price_change_24h`?

edit: okay I see that it is nominal to contrast with the percent change field which comes directly before

So instead, how about `AS price_change_24hr`, and the prior one `price_change_as_percent_24hr`

**Review by alnoki — CHANGES_REQUESTED** [review]

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 1 above)
Comment (CRBl69) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs:7]: :+1:

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 2 above)
Comment (CRBl69) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs:14]: Copy paste error.

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 3 above)
Comment (CRBl69) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs:33]: :+1:

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql (hunk 4 above)
Comment (CRBl69) [reply src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:106]: image

It looks like it gets computed once on creation.

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 5 above)
Comment (CRBl69) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs:57]: :+1:

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql (hunk 6 above)
Comment (CRBl69) [reply src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:102]: That's taken care of by the `DISTINCT ON (market_id)`, which only selects one fill event per market.

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 7 above)
Comment (CRBl69) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs:52]: No. This is unchecked. Using `query!` doesn't work here, since it doesn't look like it supports passing table/views names as an argument. And it also doesn't support passing a processed string as an argument, it wants a string literral.

File: src/rust/aggregator/src/pipelines/update_materialized_view.rs (hunk 3 above)
Comment (alnoki) [reply src/rust/aggregator/src/pipelines/update_materialized_view.rs]: Note this still says `Update...` as of 42ed45f3

I highlighted this one separately before as failing to change it wouldn't produce a compiler error

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    first_fills AS(
+        SELECT
+            *
+        FROM
+            fill_events
+        WHERE
+            "time" >= CURRENT_TIMESTAMP - '1 day'::interval
+        LIMIT 1
+    )
+    SELECT
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:58]: Now that this no longer has distinct on market ID, will the max time and first fill be correctly computed for the market ID when doing the later left join?

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+DROP VIEW api.markets;
+
+
+CREATE VIEW api.markets AS
+    WITH max_time AS (
+        SELECT
+            MAX("time") AS mtime
+        FROM
+            fill_events
+    ),
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:33]: If multiple events happen in the same txn, this will pick the earliest one, correct? And thus not necessarily the latest one in terms of emission sequence?

File: src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        LIMIT 1
+    ),
+    first_fills AS(
+        SELECT
+            *
+        FROM
+            fill_events
+        WHERE
+            "time" >= CURRENT_TIMESTAMP - '1 day'::interval
+        LIMIT 1
```
Comment (alnoki) [inline src/rust/dbv2/migrations/2023-11-16-103707_market_volume/up.sql:56]: How about pinning this to the earliest txn version/event idx?

**Review by alnoki — CHANGES_REQUESTED** [review]

**Review by alnoki — APPROVED** [review]
